mirascope/integrations/langfuse/with_langfuse.py

```diff
@@ -0,0 +1,115 @@
+"""Mirascope x Langfuse Integration."""
+
+from typing import (
+    Awaitable,
+    Callable,
+    ParamSpec,
+    TypeVar,
+)
+
+from langfuse.decorators import langfuse_context, observe
+from pydantic import BaseModel
+
+from mirascope.core.base._structured_stream import BaseStructuredStream
+
+from ...core.base import BaseCallResponse
+from ...core.base._stream import BaseStream
+from ..middleware_factory import middleware_decorator
+from ._utils import get_call_response_observation, get_stream_observation
+
+_BaseCallResponseT = TypeVar("_BaseCallResponseT", bound=BaseCallResponse)
+_BaseStreamT = TypeVar("_BaseStreamT", bound=BaseStream)
+_BaseModelT = TypeVar("_BaseModelT", bound=BaseModel)
+_BaseStructuredStreamT = TypeVar("_BaseStructuredStreamT", bound=BaseStructuredStream)
+_P = ParamSpec("_P")
+SyncFunc = Callable[
+    _P, _BaseCallResponseT | _BaseStreamT | _BaseModelT | _BaseStructuredStreamT
+]
+AsyncFunc = Callable[
+    _P,
+    Awaitable[_BaseCallResponseT | _BaseStreamT | _BaseModelT | _BaseStructuredStreamT],
+]
+_T = TypeVar("_T")
+
+
+class ModelUsage(BaseModel):
+    input: int | float | None
+    output: int | float | None
+    unit: str
+
+
+def with_langfuse(
+    fn: SyncFunc | AsyncFunc,
+) -> SyncFunc | AsyncFunc:
+    """Wraps a Mirascope function with Langfuse."""
+
+    def handle_call_response(result: BaseCallResponse, context: None):
+        langfuse_context.update_current_observation(
+            **get_call_response_observation(result, fn),
+            usage=ModelUsage(
+                input=result.input_tokens, output=result.output_tokens, unit="TOKENS"
+            ),
+        )
+
+    def handle_stream(stream: BaseStream, context: None):
+        usage = ModelUsage(
+            input=stream.input_tokens,
+            output=stream.output_tokens,
+            unit="TOKENS",
+        )
+        langfuse_context.update_current_observation(
+            **get_stream_observation(stream, fn),
+            usage=usage,
+        )
+
+    def handle_base_model(result: BaseModel | BaseStructuredStream, context: None):
```

Comment:
why not have these as two separate functions:

1. `handle_response_model`
2. `handle_structured_stream`

Also, is there a reason they should be internal functions rather than utility functions of the langfuse module that we unit test individually? That's what you've done for logfire, and I prefer that

Reply:
originally it was supposed to be because they both return a base model, but they became too different, but they also share some commonalities.

for langfuse, i'm using fn, so not sure how I would have access to fn if the function is defined outside. I would need to pass in fn as a 3rd argument to every function

Reply:
I would rather you pass in `fn` and have them as separate functions that we can unit test.

I would also rather `response_model` and `structured_stream` remain separate even if there is some commonality since those are separate paths. Even though they may share some internal properties, externally they are separate paths, and these integrations should act without any knowledge of internals